pkg/utils/k8s.go

```diff
@@ -0,0 +1,220 @@
+// if no default storage class is found.
+func GetDefaultStorageClassName() (string, error) {
+	restConfig, err := NewClientConfig()
+	if err != nil {
+		return "", err
+	}
+
+	storageV1Client, err := storagev1.NewForConfig(restConfig)
+	if err != nil {
+		return "", err
+	}
+	storageClasses, err := storageV1Client.StorageClasses().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return "", err
+	}
+	for _, storageClass := range storageClasses.Items {
+		for k, v := range storageClass.GetObjectMeta().GetAnnotations() {
+			if k == "storageclass.kubernetes.io/is-default-class" && v == "true" {
+				return storageClass.GetObjectMeta().GetName(), nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("no default storage class discovered in cluster")
+}
+
+// WaitFor runs 'fn' every 'interval' for duration of 'limit', returning no error only if 'fn' returns no
+// error inside 'limit'
+func WaitFor(fn func() error, limit, interval time.Duration) error {
+	timeout := time.After(time.Second * limit)
+	ticker := time.NewTicker(time.Second * interval)
+	defer ticker.Stop()
+	var err error
+	for {
+		select {
+		case <-timeout:
+			return fmt.Errorf("timeout with error: %v", err)
+		case <-ticker.C:
+			err = fn()
+			if err == nil {
+				return nil
+			}
+		}
+	}
+}
+
+// DeploymentIsReady attempts to `get` a deployment by name and namespace, the function returns no error
+// if no deployment replicas are ready.
+func DeploymentIsReady(config *rest.Config, name, namespace string) error {
+	clientset, err := GetClientsetFromConfig(config)
+	if err != nil {
+		return err
+	}
+	depClient := clientset.AppsV1().Deployments(namespace)
+
+	dep, err := depClient.Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if dep.Status.ReadyReplicas == 0 {
+		return fmt.Errorf("no replicas are ready for deployment %s; %s", name, namespace)
+	}
+	return nil
+}
+
+// PodIsRunning attempts to `get` a pod by name and namespace, the function returns no error
+// if the pod is in running phase.
+func PodIsRunning(config *rest.Config, name, namespace string) error {
+	clientset, err := GetClientsetFromConfig(config)
+	if err != nil {
+		return err
+	}
+	podClient := clientset.CoreV1().Pods(namespace)
+	pod, err := podClient.Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if pod.Status.Phase != corev1.PodRunning {
+		return fmt.Errorf("pod %s; %s is not in running phase", name, namespace)
+	}
+	return nil
+}
+
+// NoPodsInNS returns no error only if no pod exists in the provided namespace.
+func NoPodsInNS(config *rest.Config, namespace string) error {
+	clientset, err := GetClientsetFromConfig(config)
+	if err != nil {
+		return err
+	}
+	podClient := clientset.CoreV1().Pods(namespace)
+	pods, err := podClient.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	if len(pods.Items) > 0 {
+		return fmt.Errorf("pods still exist in namespace %s", namespace)
```

Comment:
I understand the reason of the single return value, but for me a bit annoying. The function knows about the outer world and can made the decision is it an error or not.

Reply:
I think I need to understand better why this is annoying for you 😄 Maybe we can revisit this also and you can show me a nicer way of doing it.